File: backend/src/main/java/com/_01Blog/backend/service/CommentService.java
package com._01Blog.backend.service;

import java.util.List;
import java.util.UUID;

import org.springframework.stereotype.Service;

import com._01Blog.backend.exception.ExceptionProgram;
import com._01Blog.backend.mapper.CommentMapper;
import com._01Blog.backend.model.dto.CommentDto;
import com._01Blog.backend.model.entity.Comment;
import com._01Blog.backend.model.entity.Post;
import com._01Blog.backend.model.entity.User;
import com._01Blog.backend.model.repository.CommentRepository;
import com._01Blog.backend.model.repository.PostRepository;

import jakarta.transaction.Transactional;
import lombok.NonNull;

@Service
public class CommentService {

    private CommentRepository commentRepository;
    private PostRepository postRepository;

    public CommentDto createComment(CommentDto commentDto, User user) throws ExceptionProgram {
        Post post = postRepository.findById(commentDto.getPostId())
                .orElseThrow(() -> new ExceptionProgram(400, "post not found"));
        
        if (post.isHidden()) {
            throw new ExceptionProgram(400, "you can't comment in this post");
        }

        Comment comment = new Comment();
        comment.setContent(commentDto.getContent());
        comment.setUser(user);
        comment.setPost(post);
        commentRepository.save(comment);
        return CommentMapper.toDto(comment);

        
    }

    //Get Comments
    public List<CommentDto> getComments(UUID postId, UUID userId, int offset) {
        return CommentMapper.toDto(commentRepository.getComments(postId, userId, offset));
    }

    // Delete Comment
    @Transactional
    public void deleteComment(User currentUser,@NonNull UUID commentId) throws ExceptionProgram {
        
        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new ExceptionProgram(404, "Comment not found"));

        // SECURITY CHECK: Only the owner can delete
        if (!comment.getUser().getId().equals(currentUser.getId())) {
            throw new ExceptionProgram(403, "You can only delete your own comments!");
        }

        // Optional: Also allow post owner to delete any comment
        // if (!comment.getPost().getUser().getId().equals(currentUser.getId())) {
        //     throw new ExceptionProgram(403, "Not authorized");
        // }

        commentRepository.delete(comment);
    }

}

File: frontend/src/app/features/comments/services/comment.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from '../../../core/services/api.service';
import { Observable } from 'rxjs';
import { User } from '../../../shared/models/user.model';
import { HttpParams } from '@angular/common/http';

export interface CommentDto {
    id: string; // UUID
    content: string;
    likeCount: number;
    createdAt?: string;
    user: User; // The author
    postId: string;
    likedByCurrentUser?: boolean;
}

export interface CreateCommentDto {
    content: string;
    postId: string;
}

@Injectable({
    providedIn: 'root'
})
export class CommentService {

    constructor(private api: ApiService) { }

    getComments(postId: string, offset: number = 0): Observable<CommentDto[]> {
        const params = new HttpParams()
            .set('postId', postId)
            .set('offset', offset.toString());
        return this.api.get<CommentDto[]>('get_comments', params);
    }

    createComment(comment: CreateCommentDto): Observable<CommentDto> {
        return this.api.post<CommentDto>('create_comment', comment);
    }

    deleteComment(commentId: string): Observable<any> {
        return this.api.delete(`comments/${commentId}`);
    }
}
